Objects/rangeobject.c

```diff
@@ -572,16 +572,62 @@ range_count(rangeobject *r, PyObject *ob)
 }
 
 static PyObject *
-range_index(rangeobject *r, PyObject *ob)
+range_index(rangeobject *r, PyObject *args, PyObject *kw)
 {
     int contains;
+    static char *keywords[] = {"value", "start", "stop", NULL};
+    PyObject *ob;
+    PyObject *start = NULL, *stop = NULL;
+
+    if (!PyArg_ParseTupleAndKeywords(args, kw, "O|OO:range_index",
+                                    keywords, &ob, &start, &stop)) {
+      return NULL;
+    }
+
+    if (start || stop) {
+        if (start && (!PyLong_CheckExact(start))) {
+            PyErr_Format(PyExc_TypeError,
+                         "start value must be integer, not %.200s",
+                         start->ob_type->tp_name);
+            return NULL;
+        }
+        if (stop && (!PyLong_CheckExact(stop))) {
+            PyErr_Format(PyExc_TypeError,
+                         "stop value must be integer, not %.200s",
+                         stop->ob_type->tp_name);
+            return NULL;
+        }
+
+        if (!start) {
+            start = _PyLong_GetZero();
+        }
+
+        if (!stop) {
+            stop = r->length;
+        }
+
+        PyObject *slice = PySlice_New(start, stop, _PyLong_GetOne());
+        r = (rangeobject*) compute_slice(r, slice);
```

Comment:
Note that from here `r` refers to a new object which will be leaked. But only if any of start or stop are specified.